include/revng/Support/GraphAlgorithms.h

```diff
@@ -406,138 +312,117 @@ getBackedgesBlackList(GraphT Block,
   return getBackedgesImpl<revng::detail::FilterSet::BlackList>(Block, Set);
 }
 
-// TODO: remove the `BlackList`/`WhiteList` parameter once the new AVI
-//       implementation is merged, checking that no other uses are still present
-template<revng::detail::FilterSet FilterSetType,
-         class GraphT,
-         class GT = llvm::GraphTraits<GraphT>>
-llvm::SmallSetVector<typename GT::NodeRef, 4>
-nodesBetweenImpl(GraphT Source,
-                 GraphT Target,
-                 llvm::SmallPtrSet<typename GT::NodeRef, 4> &Set) {
-  using NodeRef = typename GT::NodeRef;
-  using StateType = typename revng::detail::DFStackReachable<FilterSetType,
-                                                             NodeRef>;
-  StateType State(Set);
+template<class GraphT, class GT>
+bool hasSuccessor(GraphT Pred, GraphT Succ) {
+  for (auto Child :
+       llvm::make_range(GT::child_begin(Pred), GT::child_end(Pred))) {
+    if (Child == Succ) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/// The `nodesBetween` primitive, collects all the nodes present on any path
+/// that connects the `Source` node and the `Target` node. This function can be
+/// invoked with the second parameter, `Target`, null, and in that case it will
+/// simply collect all the nodes reachable from the `Source` node.
+template<class GraphT, class GraphDirection>
+llvm::SmallSetVector<typename llvm::GraphTraits<GraphT>::NodeRef, 4>
+nodesBetweenImpl(GraphT Source, GraphT Target) {
+  using FGT = llvm::GraphTraits<GraphDirection>;
+  using NodeRef = typename FGT::NodeRef;
 
   using SmallSetVector = llvm::SmallSetVector<NodeRef, 4>;
+  SmallSetVector ForwardDFSNodes;
+  SmallSetVector BackwardDFSNodes;
 
-  // Assign the `Source` node
-  State.assignSource(Source);
+  // We cannot perform a `nodesBetween` search starting from a null `Source`
+  revng_assert(Source != nullptr);
 
-  // Initialize the visited set with the target node, which is the boundary
-  // that we don't want to trepass when finding reachable nodes
-  State.assignTarget(Target);
-  State.insertTarget(Target);
-  State.insertInMap(Target, false);
+  // Perform the forward DFS visit, stopping at the `Target` node
+  using ExtType = typename llvm::df_iterator_default_set<NodeRef>;
+  ExtType ForwardExtSet;
+  ForwardExtSet.insert(Target);
 
-  using nbdf_iterator = llvm::df_iterator<GraphT, StateType, true, GT>;
-  auto Begin = nbdf_iterator::begin(Source, State);
-  auto End = nbdf_iterator::end(Source, State);
+  using fdf_iterator = llvm::df_iterator<GraphT, ExtType, true, FGT>;
+  auto ForwardBegin = fdf_iterator::begin(Source, ForwardExtSet);
+  auto ForwardEnd = fdf_iterator::end(Source, ForwardExtSet);
 
-  for (NodeRef Block : llvm::make_range(Begin, End)) {
-    (void) Block;
+  for (NodeRef Block : llvm::make_range(ForwardBegin, ForwardEnd)) {
+    ForwardDFSNodes.insert(Block);
```

Comment:
Any reason you need `ForwardDFSNodes` instead of using `ForwardExtSet` directly?

Reply:
This is a very good question, we discussed a bit over it, and the final decision was to leave it as it is 9b5a0ed0e76593d28bf4787c768a11739a7f2469